docs/internal/3i.md

```diff
@@ -0,0 +1,141 @@
+
+The cleanup process must complete system-level invariants such as: unmapping vmmap regions, cleaning fdtables, waking waiters, canceling schedulers or timers, etc.. Allowing these calls to be interposable would permit third-party grates to inject arbitrary logic (e.g., blocking, allocation, or reentrancy), which could stall or disrupt the teardown sequence, resulting in resource leaks, deadlocks, or zombie cages/grates.
+
+## Grates
+
+### Overview
+
+Grates are cages that provide services to descendant cages by interposing on their system calls. They can execute trusted operations (e.g., syscall filtering, tracing, or in-memory file I/O) on behalf of other cages without requiring changes to the microvisor’s trusted computing base (TCB).
+
+From 3i’s perspective, a grate is both:
+
+- A consumer of system calls from descendant cages, and
+- A service provider exposing its own callable entry points to the 3i runtime.
+
+### Design Goals
+
+3i’s public API is runtime-agnostic, allowing it to be adapted to multiple backends (e.g., Wasmtime, MPK, or future runtimes).
+
+Runtime-specific mechanisms, such as Wasmtime’s `VMContext` management, are implemented in the corresponding runtime layer (e.g., the Wasmtime adapter). These runtime hooks are exposed to 3i as C-ABI function pointers and registered in 3i’s global table, decoupling runtime details from 3i’s core logic.
+
+### Wasm Implementation
+
+Grate execution relies on a re-entry mechanism that bridges Wasm runtime contexts (e.g., Wasmtime instances) and 3i’s routing layer. 
+
+Each Wasm module in Wasmtime has its own `VMContext`, which encapsulates all runtime state required for that instance’s execution. When 3i performs a cross-module call (e.g., from _cage A_ to a _grate B_), the currently executing context belongs to _A_, but 3i needs to restore the latest runtime state of _B_ before re-entering its Wasm code. 
+
+Because Rust’s lifetime system prevents us from directly sharing or globally storing references to the caller’s runtime structures, we instead store and pass raw pointers to the target module’s `VMContext`. 3i uses a registered callback function pointer and retrived the corresponding runtime context (`VMContext`) from pointer to re-enter the target Wasm instance. This allows 3i to reconstruct the correct execution environment for _B_ at re-entry time, ensuring cross-instance transitions without violating Rust’s ownership and lifetime rules.
+
+During module initialization, Wasmtime captures each instance’s low-level `VMContext`, representing its active execution state (memories, globals, tables, etc.). However, the registration of interposition handlers (`register_handler`) often occurs after initialization. To bridge this timing gap, Wasmtime maintains a runtime-local table:
```

Comment:
I think that 3i will contain the information (in the optional info of the register_handler / make_syscall calls) to handle this.  Do you need a table still?

Reply:
Sorry, this only contains the jump address.  If the tid / pid (cageid) are consistent, then I think you could look this up in the cases you need the context.